File: ts/project/.sdk/tm/php/utility/TransformRequest.php
<?php
declare(strict_types=1);

// ProjectName SDK utility: transform_request

require_once __DIR__ . '/../core/Helpers.php';

class ProjectNameTransformRequest
{
    public static function call(ProjectNameContext $ctx): mixed
    {
        $spec = $ctx->spec;
        $point = $ctx->point;
        if ($spec) {
            $spec->step = 'reqform';
        }
        $transform = ProjectNameHelpers::to_map(\Voxgig\Struct\Struct::getprop($point, 'transform'));
        if (!$transform) {
            return self::strip_action($ctx->reqdata);
        }
        $reqform = \Voxgig\Struct\Struct::getprop($transform, 'req');
        if (!$reqform) {
            return self::strip_action($ctx->reqdata);
        }
        return self::strip_action(\Voxgig\Struct\Struct::transform(['reqdata' => $ctx->reqdata], $reqform));
    }

    // `$action` selects the point (see MakePoint); it is never an API field,
    // so the body is a copy without it. Arrays are values, so the unset
    // never reaches the caller's copy.
    private static function strip_action(mixed $reqdata): mixed
    {
        if (!is_array($reqdata) || !array_key_exists('$action', $reqdata)) {
            return $reqdata;
        }
        unset($reqdata['$action']);
        return $reqdata;
    }
}

File: ts/project/.sdk/tm/rb/utility/transform_request.rb
# ProjectName SDK utility: transform_request
require_relative 'struct/voxgig_struct'
require_relative '../core/helpers'
module ProjectNameUtilities
  # `$action` selects the point (see MakePoint); it is never an API field, so
  # the body is a copy without it. The caller's hash is left untouched.
  def self.strip_action(reqdata)
    return reqdata unless reqdata.is_a?(Hash) && reqdata.key?("$action")
    reqdata.reject { |k, _| k == "$action" }
  end

  TransformRequest = ->(ctx) {
    spec = ctx.spec
    point = ctx.point
    spec.step = "reqform" if spec
    transform = ProjectNameHelpers.to_map(VoxgigStruct.getprop(point, "transform"))
    return ProjectNameUtilities.strip_action(ctx.reqdata) unless transform
    reqform = VoxgigStruct.getprop(transform, "req")
    return ProjectNameUtilities.strip_action(ctx.reqdata) unless reqform
    ProjectNameUtilities.strip_action(VoxgigStruct.transform({ "reqdata" => ctx.reqdata }, reqform))
  }
end
